processing/src/main/java/org/apache/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -1038,4 +1204,410 @@ public void clear()
+    {
+      return aggregators.get(rowOffset)[aggOffset].isNull();
+    }
+
+    @Nullable
+    @Override
+    public ColumnCapabilities getColumnCapabilities(String column)
+    {
+      if (ColumnHolder.TIME_COLUMN_NAME.equals(column) || Objects.equals(column, timeColumnName)) {
+        return ColumnCapabilitiesImpl.createDefault().setType(ColumnType.LONG).setHasNulls(false);
+      }
+      if (dimensionsMap.containsKey(column)) {
+        return dimensionsMap.get(column).getCapabilities();
+      }
+      if (aggregatorsMap.containsKey(column)) {
+        return aggregatorsMap.get(column).getCapabilities();
+      }
+      return null;
+    }
+
+    /**
+     * Ensure that the projection has the specified column required by a {@link CursorBuildSpec} in one form or another.
+     * If the column is a {@link VirtualColumn} on the build spec, ensure that the projection has an equivalent virtual
+     * column, or has the required inputs to compute the virtual column. If an equivalent virtual column exists, its
+     * name will be added to the 'rewriteVirtualColumns' parameter  so that the build spec name can be mapped to the
+     * projection name. If no equivalent virtual column exists, but the inputs are available on the projection to
+     * compute it, it will be added to 'referencedVirtualColumns' parameter.
+     */
+    private boolean hasRequiredColumn(
+        String column,
+        VirtualColumns buildSpecVirtualColumns,
+        Set<VirtualColumn> referencedVirtualColumns,
+        Map<String, String> rewriteVirtualColumns
+    )
+    {
+      final VirtualColumn buildSpecVirtualColumn = buildSpecVirtualColumns.getVirtualColumn(column);
+      if (buildSpecVirtualColumn != null) {
+        // check to see if we have an equivalent virtual column defined in the projection, if so we can
+        final VirtualColumn projectionEquivalent = virtualColumns.findEquivalent(buildSpecVirtualColumn);
+        if (projectionEquivalent != null) {
+          if (!buildSpecVirtualColumn.getOutputName().equals(projectionEquivalent.getOutputName())) {
+            rewriteVirtualColumns.put(buildSpecVirtualColumn.getOutputName(), projectionEquivalent.getOutputName());
+          }
+          return true;
+        }
+
+        referencedVirtualColumns.add(buildSpecVirtualColumn);
+        final List<String> requiredInputs = buildSpecVirtualColumn.requiredColumns();
+        if (requiredInputs.size() == 1 && ColumnHolder.TIME_COLUMN_NAME.equals(requiredInputs.get(0))) {
+          // special handle time granularity
+          final Granularity virtualGranularity = Granularities.fromVirtualColumn(buildSpecVirtualColumn);
+          if (virtualGranularity != null) {
+            return !virtualGranularity.isFinerThan(projectionGranularity);
+          } else {
+            // anything else with __time requires none granularity
+            return Granularities.NONE.equals(projectionGranularity);
+          }
```

Comment:
how does it work if requiredInputs are more than 1 and one of the required input is a time column?